### ISSUE TITLE
"[CS8603] Possible null reference return" emitted even though NotNullIfNotNull used.

### ISSUE THREAD

**Issue by jahmai** [dotnet/roslyn#39728]
**Version Used**: 

VS 16.3.8
Netstandard 2.1

**Steps to Reproduce**:

1.
Create a new .netstandard 2.1 project, add `<Nullable>enable</Nullable>` and `<LangVersion>8.0</LangVersion>` to the project file

2.

Add the following class:

```
    public sealed class Bytes
    {
        public static readonly Bytes Empty = new Bytes();

        [return: NotNullIfNotNull("left")]
        [return: NotNullIfNotNull("right")]
        public static Bytes? operator +(Bytes? left, Bytes? right) => left == null ? right : right == null ? left : Empty;

        public static Bytes Example()
        {
            return Empty + Empty;
        }
    }

```

3. Compile

**Expected Behavior**:

No warnings emitted

**Actual Behavior**:

"[CS8603] Possible null reference return" emitted for the method Example()

**Comment by jahmai** [comment]
Similarly with cast operators:

```
    public class Class1
    {
        public XElement Element { get; set; } = new XElement(XNamespace.None.GetName("eh"));

        [return: NotNullIfNotNull("value")]
        public static explicit operator XElement?(Class1? value) => value?.Element;

        public void Cast()
        {
            var fragment = new Class1();
            var element = (XElement) fragment; //  [CS8600] Converting null literal or possible null value to non-nullable type.
            Class1? fragment2 = null;
            var element2 = (XElement?) fragment2;
        }
    }
```

I expect that because "fragment" above is non-nullable, and the cast operator has the NotNullIfNotNull attribute, that it wouldn't complain because the return value should be non-null.

**Comment by jahmai** [comment]
Duplicate of #37870 #32671 #38132
